engine/src/main/scala/cromwell/engine/workflow/lifecycle/execution/callcaching/EngineJobHashingActor.scala

```diff
@@ -1,53 +1,112 @@
+      // Also update the state data with these new hash results.
+      val newHashResults = hashResultMessage.hashes
+      findCacheResults(newHashResults)
+      stay using updateStateDataWithNewHashResults(newHashResults)
+    case Event(newCacheResults: CacheResultMatchesForHashes, _) =>
+      checkWhetherHitOrMissIsKnownThenTransition(stateData.intersectCacheResults(newCacheResults))
+    case Event(PlaceholderHashKeyExpansion(placeholderHashKey, newHashKeys), _) => checkWhetherHitOrMissIsKnownThenTransition(stateData.replacePlaceholderHashKey(placeholderHashKey, newHashKeys))
   }
 
   when(GeneratingAllHashes) {
-    case Event(hashResultMessage: HashResultMessage, _) => checkWhetherAllHashesAreKnownAndTransition(stateData.withNewKnownHashes(hashResultMessage.hashes))
+    case Event(hashResultMessage: SuccessfulHashResultMessage, _) => checkWhetherAllHashesAreKnownAndTransition(stateData.withNewKnownHashes(hashResultMessage.hashes))
+    case Event(PlaceholderHashKeyExpansion(placeholderHashKey, newHashKeys), _) => checkWhetherAllHashesAreKnownAndTransition(stateData.replacePlaceholderHashKey(placeholderHashKey, newHashKeys))
     case Event(CacheResultMatchesForHashes(_, _), _) => stay // Don't care; we already know the hit/miss status. Ignore this message
   }
 
+  whenUnhandled {
+    case Event(failure: CacheResultLookupFailure, _) =>
+      // Crash and let the supervisor deal with it.
+      throw new RuntimeException("Failure looking up call cache results", failure.reason)
+    case Event(HashingFailedMessage(hashKey, reason), _) =>
+      receiver ! HashError(new Exception(s"Unable to generate ${hashKey.key} hash. Caused by ${reason.getMessage}", reason))
+      context.stop(self)
+      stay
+  }
+
   private def initializeEJHA() = {
 
-    // TODO: Implement call caching hashing: #1230
-    val initialHashes = List(
-      HashResult(HashKey("commandTemplate"), HashValue(jobDescriptor.call.task.commandTemplate.toString))
-    )
-    val fileHashesNeeded: Map[HashKey, Path] = Map.empty
-    val hashesNeeded: List[HashKey] = initialHashes.map(_.hashKey) ++ fileHashesNeeded.keys
+    import cromwell.core.simpleton.WdlValueSimpleton._
+    val inputSimpletons = jobDescriptor.inputs.simplify
+    val (fileInputSimpletons, nonFileInputSimpletons) = inputSimpletons partition {
+      case WdlValueSimpleton(_, f: WdlFile) => true
+      case _ => false
+    }
+
+    val initialHashes = calculateInitialHashes(nonFileInputSimpletons, fileInputSimpletons)
+
+    val fileContentHashesNeeded = if (mode.hashFileContents) {
+      JobFileHashRequests(jobDescriptor.key, fileInputSimpletons map { case WdlValueSimpleton(name, x: WdlFile) => SingleFileHashRequest(HashKey(s"input: File(Content) $name"), x) })
+    } else {
+      JobFileHashRequests(jobDescriptor.key, List.empty)
+    }
+    val runtimeAttributeHashesNeededPlaceholder = HashKey("PLACEHOLDER: runtime attributes")
```

Comment:
Is there value in having an explicit placeholder `HashKey`?  It seems like this just travels from here to the backend specific hasher actor and back.  It gets added to `hashesNeeded` below but then filtered out.